File: solutions/43.cpp
#include <iostream>
#include <algorithm>
#include <string>
#include <vector>
#include <sstream>
#include <iterator>
#include <set>
#include <stack>
#include <numeric>
#include <regex>
#include <queue>
#include <iomanip>
using namespace std;

#define FILE_IO {freopen("input.txt", "r", stdin); freopen("output.txt", "w", stdout);}
#define FAST_IO ios_base::sync_with_stdio(0); cin.tie(NULL), cout.tie(NULL)

#define id first
#define vl second

void dfs(int u, vector<vector<int>> &G, vector<bool> &visited) {
    visited[u] = true;
    for (int v : G[u]) {
        if (!visited[v]) {
            dfs(v, G, visited);
        }
    }
}

int main() {
    FAST_IO;
    // FILE_IO;

    vector<vector<int>> G(100010);
    vector<bool> vst(100010);
    int T, N, E, u, v;
    cin >> T;

    while (T--) {
        cin >> N >> E;
        for (int i = 0; i < N; ++i){
            G[i].clear();
            vst[i] = false;
        }

        for (int i = 0; i < E; ++i) {
            cin >> u >> v;
            G[u].push_back(v);
            G[v].push_back(u);
        }

        int cnt = 0;
        for (int i = 0; i < N; ++i) {
            if (!vst[i]) {
                dfs(i, G, vst);
                ++cnt;
            }
        }

        cout << cnt << endl;
    }


    return 0;
}

File: 43.py
import sys
from typing import List


def FILE_IO():
    sys.stdin = open('input.txt', 'r')
    sys.stdout = open('output.txt', 'w')
    
def dfs(G: List[List[int]], vst: List[bool], u: int):
    vst[u] = True
    for v in G[u]:
        if not vst[v]:
            dfs(G, vst, v)

def solve():
    T = int(input())
    G: List[List[int]] = [[] for _ in range(100_005)]
    vst: List[bool] = [False for _ in range(100_005)]
    for _ in range(T):
        input()
        N = int(input())
        E = int(input())
        for i in range(N):
            G[i].clear()
            vst[i] = False
        
        for _ in range(E):
            u, v = map(int, input().split())
            G[u].append(v)
            G[v].append(u)
            
        cnt = 0
        for i in range(N):
            if not vst[i]:
                cnt += 1
                dfs(G, vst, i)
        
        print(cnt)
        
# FILE_IO()
solve()